File: go/keymanager/churp/policy.go
package churp

import (
	"fmt"

	"github.com/oasisprotocol/oasis-core/go/common/cbor"
	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
	"github.com/oasisprotocol/oasis-core/go/common/sgx"
)

// PolicySGXSignatureContext is the context used to sign PolicySGX documents.
var PolicySGXSignatureContext = signature.NewContext("oasis-core/keymanager/churp: policy")

// PolicySGX represents an SGX access control policy used to authenticate
// key manager enclaves during handoffs.
type PolicySGX struct {
	Identity

	// Serial is the monotonically increasing policy serial number.
	Serial uint32 `json:"serial"`

	// MayShare is the vector of enclave identities from which a share can be
	// obtained during handouts.
	MayShare []sgx.EnclaveIdentity `json:"may_share"`

	// MayJoin is the vector of enclave identities that may form the new
	// committee in the next handoffs.
	MayJoin []sgx.EnclaveIdentity `json:"may_join"`
}

// SanityCheck verifies the validity of the policy.
func (p *PolicySGX) SanityCheck(prev *PolicySGX) error {
	switch prev {
	case nil:
		switch {
		case p.Serial != 0:
			return fmt.Errorf("SGX policy: sanity check failed: serial number must be 0 and not %d", p.Serial)
		}
	default:
		switch {
		case p.ID != prev.ID:
			return fmt.Errorf("SGX policy: sanity check failed: ID changed from %d to %d", prev.ID, p.ID)
		case p.RuntimeID != prev.RuntimeID:
			return fmt.Errorf("SGX policy: sanity check failed: runtime ID changed from %s to %s", prev.RuntimeID, p.RuntimeID)
		case p.Serial != prev.Serial+1:
			return fmt.Errorf("SGX policy: sanity check failed: serial number must be %d and not %d", prev.Serial+1, p.Serial)
		}
	}

	return nil
}

// SignedPolicySGX represents a signed SGX access control policy.
//
// The runtime extension will accept the policy only if all signatures are
// valid, and a sufficient number of trusted policy signers have signed it.
type SignedPolicySGX struct {
	// Policy is an SGX access control policy.
	Policy PolicySGX `json:"policy"`

	// Signatures is a vector of signatures.
	Signatures []signature.Signature `json:"signatures,omitempty"`
}

// SanityCheck verifies the validity of the policy and the signatures.
func (p *SignedPolicySGX) SanityCheck(prev *PolicySGX) error {
	if err := p.Policy.SanityCheck(prev); err != nil {
		return err
	}

	raw := cbor.Marshal(p.Policy)
	for _, sig := range p.Signatures {
		if !sig.PublicKey.IsValid() {
			return fmt.Errorf("SGX policy: sanity check failed: signature's public key %s is invalid", sig.PublicKey.String())
		}
		if !sig.Verify(PolicySGXSignatureContext, raw) {
			return fmt.Errorf("SGX policy: sanity check failed: policy signature from %s is invalid", sig.PublicKey.String())
		}
	}

	return nil
}

// Sign signs the policy with the given signer and appends the signatures.
func (p *SignedPolicySGX) Sign(signers []signature.Signer) error {
	if len(signers) == 0 {
		return nil
	}

	rawPolicy := cbor.Marshal(p.Policy)

	sigs := make([]signature.Signature, 0, len(signers))
	for _, signer := range signers {
		sig, err := signature.Sign(signer, PolicySGXSignatureContext, rawPolicy)
		if err != nil {
			return err
		}
		sigs = append(sigs, *sig)
	}

	p.Signatures = append(p.Signatures, sigs...)

	return nil
}

File: keymanager/src/churp/policy.rs
//! Policy support.
use std::{
    cmp::Ordering,
    collections::{HashMap, HashSet},
    iter::FromIterator,
    sync::{Arc, Mutex},
};

use anyhow::Result;

use oasis_core_runtime::{
    common::{namespace::Namespace, sgx::EnclaveIdentity},
    consensus::keymanager::churp::{PolicySGX, SignedPolicySGX},
};

use crate::policy::verify_data_and_trusted_signers;

use super::Error;

/// A collection of cached key manager policies.
pub struct VerifiedPolicies {
    policies: Mutex<HashMap<u8, Arc<VerifiedPolicy>>>,
}

impl VerifiedPolicies {
    /// Creates a new collection of cached verified policies.
    pub fn new() -> Self {
        Self {
            policies: Mutex::new(HashMap::new()),
        }
    }

    /// Verifies and caches the given policy.
    pub fn verify(&self, policy: &SignedPolicySGX) -> Result<Arc<VerifiedPolicy>> {
        let mut policies = self.policies.lock().unwrap();

        let churp_id = policy.policy.id;
        let verified_policy = policies.get(&churp_id);
        if let Some(verified_policy) = verified_policy {
            match verified_policy.serial.cmp(&policy.policy.serial) {
                Ordering::Less => (),
                Ordering::Equal => return Ok(verified_policy.clone()),
                Ordering::Greater => return Err(Error::PolicyRollback.into()),
            }
        }

        let verified_policy = verify_data_and_trusted_signers(policy)?;
        let verified_policy = VerifiedPolicy::new(verified_policy)?;
        let verified_policy = Arc::new(verified_policy);
        policies.insert(churp_id, verified_policy.clone());

        Ok(verified_policy)
    }
}

impl Default for VerifiedPolicies {
    fn default() -> Self {
        Self::new()
    }
}

/// A policy verified to be signed by trusted signers and published
/// in the consensus layer.
#[derive(Clone, Default, Debug)]
pub struct VerifiedPolicy {
    /// A monotonically increasing policy serial number.
    pub serial: u32,

    /// A set of enclave identities from which a share can be obtained
    /// during handouts.
    pub may_share: HashSet<EnclaveIdentity>,

    /// A set of enclave identities that may form the new committee
    /// in the next handoffs.
    pub may_join: HashSet<EnclaveIdentity>,

    /// A map of runtime identities and their respective sets of enclave
    /// identities that are allowed to query key shares.
    pub may_query: HashMap<Namespace, HashSet<EnclaveIdentity>>,
}

impl VerifiedPolicy {
    /// Creates a new policy from the given SGX policy.
    ///
    /// The provided policy should be valid, signed by trusted signers,
    /// and published in the consensus layer state.
    fn new(verified_policy: &PolicySGX) -> Result<Self> {
        let may_share = HashSet::from_iter(verified_policy.may_share.iter().cloned());
        let may_join = HashSet::from_iter(verified_policy.may_join.iter().cloned());
        let may_query = HashMap::from_iter(verified_policy.may_query.iter().map(
            |(runtime_id, enclaves)| (*runtime_id, HashSet::from_iter(enclaves.iter().cloned())),
        ));

        Ok(Self {
            serial: verified_policy.serial,
            may_share,
            may_join,
            may_query,
        })
    }

    /// Returns true iff shares can be obtained from the remote enclave
    /// during handouts.
    pub fn may_share(&self, remote_enclave: &EnclaveIdentity) -> bool {
        self.may_share.contains(remote_enclave)
    }

    /// Returns true iff the remote enclave is allowed to form the new
    /// committee in the next handoffs.
    pub fn may_join(&self, remote_enclave: &EnclaveIdentity) -> bool {
        self.may_join.contains(remote_enclave)
    }

    /// Returns true iff the remote enclave is allowed to query key shares
    /// for the given runtime.
    pub fn may_query(&self, remote_enclave: &EnclaveIdentity, runtime_id: &Namespace) -> bool {
        self.may_query
            .get(runtime_id)
            .map(|may_query_runtime| may_query_runtime.contains(remote_enclave))
            .unwrap_or(false)
    }
}
